### ISSUE TITLE
Data Table Enhancement: Compound Expansion 

### ISSUE THREAD

**Issue by LHinson** [patternfly/patternfly-next#721]
Enhance the Data Table to support compound expansion. 

List + Specs: https://drive.google.com/drive/u/1/folders/1cKk-HnEJFcB4w6sgl05NeLPVLRUUiNUW

Related issue #198

Component lifecycle: 
- [x] Interaction design requirements 
- [x] Visuals 
- [x] First checkpoint complete and approved 
- [ ] CSS 

[Compound Expansion - Expansion Hover.pdf](https://github.com/patternfly/patternfly-next/files/2650769/Compound.Expansion.-.Expansion.Hover.pdf)
[compound-expansion-mobile.pdf](https://github.com/patternfly/patternfly-next/files/2650770/compound-expansion-mobile.pdf)

**Comment by maryshak1996** [comment]
@LHinson Kyle wants me to work on this next, but I still don't have the ability to assign things to myself -- would you mind assigning to me when you get a chance? :)

**Comment by LHinson** [comment]
you got it! @maryshak1996

**Comment by maryshak1996** [comment]
Some communication about the visuals:

@andresgalante  - "I think we probably need an active state to show user which element is active. Having said that, on the implementation part this is a button inside a table.
Let me explain: we have Data table. Data table can help any content inside, it can be text, an image, an icon, a button or even another table.
I suggest we don't change the styles of a button just because they are in the table. We have link button that looks like the ones you draw."

@mcarrano - "I agree with Andres' comment. Also, is it assumed that we can support this capability for either a data list or a data table? Seems like we should."

**Comment by maryshak1996** [comment]
@andresgalante : the idea that Kyle and I came up with, that I'm trying to get across here, is that the active state is indicated by the grey background appearing behind the active link and the link itself becoming thicker in weight. This is the same treatment as vertical subnav elements.

**Comment by maryshak1996** [comment]
![compound-expansion_checkpoint1](https://user-images.githubusercontent.com/19824574/46555051-2841cb80-c8b0-11e8-8adc-10dc6dd150d7.png)
Above comments were in reference to this design

**Comment by mcarrano** [comment]
OK, I see @maryshak1996 . This could work. I'm also wondering whether we should use a similar visual approach that is being used for tabs (i.e. to show the selected tab). Have you seen that?

**Comment by maryshak1996** [comment]
@mcarrano thanks for that idea - I gave it a try and I think that it works well 
![compound-expansion_blue-selected-outline](https://user-images.githubusercontent.com/19824574/46692955-918a4d00-cbd6-11e8-960d-81c26bd70901.png)

**Comment by mcarrano** [comment]
I like it as well @maryshak1996. Will be interesting to hear what others think.

**Comment by maryshak1996** [comment]
@kybaker what do you think of that treatment of the selected state?

**Comment by kybaker** [comment]
@maryshak1996 I like this solution. Though there are a few places that are inconsistent in the table. The shadows are missing and the kebab and arrow icons are inconsistent. Need to make these updates.

**Comment by maryshak1996** [comment]
@mcarrano @andresgalante 
I worked with kyle to finalize the visuals for this, so here is a quick screenshot but you can also look at the invision link for it as well
https://redhat.invisionapp.com/d/main#/console/15596505/324619680/preview
![compound-expansion-final-design](https://user-images.githubusercontent.com/19824574/46748934-c05dfd00-cc82-11e8-94f3-394fd7e2151b.png)

**Comment by mcarrano** [comment]
I think this looks great @maryshak1996 . The only small thing I would point out is that the alignment of the width of the "tab" with the heading "Branches" seems unexpected. Since this is a table, my expectation is that this would be implemented by placing a border on the table cell (@andresgalante do you agree?). Therefore I would expect the width of the tab to match the width of the column and the column heading to fit in that same space. Does that make sense?

**Comment by mcarrano** [comment]
@LHinson a question about this issue. The title references "Data List expansion" but the mockups are really for Data Table expansion. Should the title be changed to reflect this? Do we really need or want list expansion? both?

The way we are defining data lists for PF4, I'm not sure that the compound expansion use case makes sense and based on what @maryshak1996 designed, IMO this should be Data Table Expansion but wanted to check before editing.

**Comment by maryshak1996** [comment]
@mcarrano Totally see where you're at with the alignment of the selected state - here's my solution:
![screen shot 2018-10-18 at 4 48 19 pm](https://user-images.githubusercontent.com/19824574/47183326-bc228700-d2f5-11e8-97dc-91b193a4cd9a.png)
In order for the grey bars on the left and right of the selected item to be present, we need to create some margin (16px to be exact) to the left of the header items.

**Comment by mcarrano** [comment]
This is perfect! Thanks @maryshak1996

**Comment by rachael-phillips** [comment]
Referencing Christie's Data Table Enhancement: Compound Expansion #970 to close this issue

**Comment by rachael-phillips** [comment]
Sorry to cause any confusion! Reopening this!

**Comment by mcoker** [comment]
Attaching @maryshak1996's design for mobile:

![screen shot 2019-01-09 at 11 47 14 am](https://user-images.githubusercontent.com/35148959/50914510-4dd4f880-13fc-11e9-859e-ceb606236bcc.png)

**Comment by maryshak1996** [comment]
@mcoker @christiemolloy here's a mockup that reflects that we talked about this morning :) 
![compound-expansion-mobile](https://user-images.githubusercontent.com/19824574/51272787-1684bf80-1999-11e9-9e47-d966372d0cf0.png)
And here are the specs (a full PDF spec of the full table and a screenshot showing a close-up for the spacing around the expanded item/blue underline treatment)
![zoom-in](https://user-images.githubusercontent.com/19824574/51272843-3ddb8c80-1999-11e9-87df-4c4b10a913d0.png)
[compound-expansion-mobile_spec.pdf](https://github.com/patternfly/patternfly-next/files/2765702/compound-expansion-mobile_spec.pdf)

**Comment by mcoker** [comment]
Should the grey divider under description, date, and status be full-width? There is a left/right gap in the design.

<img width="969" alt="screen shot 2019-01-23 at 9 06 00 pm" src="https://user-images.githubusercontent.com/35148959/51651789-e453fe80-1f52-11e9-8a85-3886f86884ab.png">

Should the grey divider be visible here? Not seeing it in the design.
<img width="526" alt="screen shot 2019-01-23 at 9 12 17 pm" src="https://user-images.githubusercontent.com/35148959/51651998-aacfc300-1f53-11e9-89b7-45d7a01481dd.png">

The scrollbar is disabled in expanded rows - is that on purpose? I didn't know I could scroll until I accidentally scrolled.

I'm seeing this weird background-color and it's not obvious where it's coming from.
<img width="785" alt="screen shot 2019-01-23 at 9 25 18 pm" src="https://user-images.githubusercontent.com/35148959/51652508-896fd680-1f55-11e9-9333-d250c796810a.png">

There is something weird with the first expandable column in the compound expansion rows ("Repository Name"). In Chrome, if you hover over that text, you get a normal cursor. But if you hover over the text of any of the other rows (Branches, Pull Requests, Work Spaces), you get a pointer cursor. I'm assuming that's coming from the `::after` pseudo element because using a negative `z-index` disables that behavior. It's weird that you can click on the text (in the mobile/responsive view)  the button has a focus ring, but the button is beside the text and not where you clicked.

**Comment by mcarrano** [comment]
@maryshak1996 @gdoyle1 Can one of you try to answer this? @mcoker Is there an open PR/preview we can look at? Not seeing a link.

**Comment by maryshak1996** [comment]
@mcoker to address your design questions:

> Should the grey divider under description, date, and status be full-width? There is a left/right gap in the design.
Yes, the divider should be full-width -- my bad, I made this update on my end but forgot to update here

> Should the grey divider be visible here? Not seeing it in the design.
I would prefer to not have the divider here, as I think it creates some confusion that that inset/expanded table content is part of the another table item
 

> The scrollbar is disabled in expanded rows - is that on purpose? I didn't know I could scroll until I accidentally scrolled.
That's a good point. I know that Christie and I discussed having a max height on the expansion, but I don't remember discussions around the scrollbar. Could we have the scroll bar appear just when the expanded content exceeds that max height?
